lib/api/websiteGet.js

```diff
@@ -138,6 +138,10 @@ function websiteGet(request, log, callback) {
                     log.trace('error retrieving object metadata',
                     { error: err });
                     let returnErr = err;
+                    // requestType must be bucketGet to check bucket acl
+                    // but objectGet to check bucket policy
+                    // eslint-disable-next-line no-param-reassign
+                    request.websiteOverrideRequestTypeForBucketPolicy = true;
```

Comment:
As this is API-specific logic, would it work better if we handle the return code directly in the API, in case of an access denied?
Based on the authz results from IAM,overriding the request type may cause corner cases where some actions are allowed or not by the ACLs and bucket policy. I think, if we need to check both bucket ACLs and ObjectGet permission, that we should call this function twice and behave like AWS.

In websiteGet, we already check the bucket ACLs and policy with the `bucketGet` request type in case we have an error getting the object MD:
```js
// if index object does not exist and bucket is private AWS
                    // returns 403 - AccessDenied error.
                    if (err.is.NoSuchKey && !bucketAuthorized) {
                        returnErr = errors.AccessDenied;
                    }
                    return _errorActions(returnErr,
                      websiteConfig.getErrorDocument(), routingRules,
                      bucket, reqObjectKey, corsHeaders, request, log,
                      callback);
```
Here, we should be able to either evaluate the `objectGet` permission to return 404. We have the standardMetadataValidateBucketAndObj function that can be used, and from which you can override the request type, without altering the request object, eg:

```js
    const mdValParams = {
        authInfo,
        bucketName,
        objectKey,
        versionId,
        getDeleteMarker: true,
        requestType: 'objectHead',
        request,
    };

    return standardMetadataValidateBucketAndObj(mdValParams, request.actionImplicitDenies, log, [...]
```

Reply:
I can't run it twice, all these functions seems to encapsulate both `checkBucketAcls` and `processBucketPolicy` but I need to run them separately for `bucketGet` and `objectGet`.

`processBucketPolicy` with `bucketGet` is invalid because website only accepts `s3:GetObject` so if there is `s3:ListBucket` in a policy this will result in an invalid access.

`checkBucketAcls` returns always `true` for `objectGet` (to delegate to object acl, but I'm in the case where there is no object)

Reply:
The idea is to have an additional permission check, only if the object is not found but we are supposed to have access to it (given that it exists), in this case, we should check the policies against `s3:ObjectGet` instead of `s3:bucketGet`.
If the `isBucketAuthorized` or `isObjAuthorized` are checking the ACLs but you want to skip them, shouldn't we just update these functions with a new optional parameter so that we skip the ACL checks, to avoid changing the generic function?

Reply:
Yes but I didn't wanted to add another parameter as theses functions already have a lot of parameters.
But I can do that so I don't put a flag on the request and I can override the request type earlier